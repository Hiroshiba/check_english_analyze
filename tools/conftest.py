import pytest
from syrupy.assertion import SnapshotAssertion
from syrupy.extensions.json import JSONSnapshotExtension


@pytest.fixture
def snapshot_json(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    """JSON形式のスナップショットテスト用fixture"""
    return snapshot.use_extension(JSONSnapshotExtension)
