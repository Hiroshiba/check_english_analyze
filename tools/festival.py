import argparse
import json
import subprocess
from pathlib import Path

import sexpdata
from pydantic import BaseModel

from utility.logger_utility import get_logger, logging_setting

logging_setting()
logger = get_logger(Path(__file__))


class PhonemeInfo(BaseModel):
    """単語・シラブル・音素・ストレス情報"""

    word: str
    syllable_index: int | None
    phoneme: str | None
    stress: int | None


def main():
    """英語テキストから音素・シラブル・ストレス有無を抽出しprintする"""
    text = parse_args()
    script = build_festival_script(text)
    output = run_festival(script)
    infos = extract_sexp(output)
    print_phoneme_info(infos)


def parse_args() -> str:
    """コマンドライン引数からテキストを取得する"""
    parser = argparse.ArgumentParser()
    parser.add_argument("text", type=str, help="解析するテキスト")
    args = parser.parse_args()
    return args.text


def build_festival_script(text: str) -> str:
    """Festival用Schemeスクリプトを生成する"""
    return f"""(voice_cmu_us_slt_arctic_clunits)
(lex.select "cmu")

(define (phonemize line)
  (set! utterance (eval (list 'Utterance 'Text line)))
  (utt.synth utterance)
  (print (utt.relation_tree utterance "SylStructure")))

(set! lines (list "{text}"))
(mapcar (lambda (line) (phonemize line)) lines)
"""


def run_festival(script: str) -> str:
    """Festivalを実行し出力を得る"""
    print(f"script: {script}")
    try:
        output = subprocess.check_output(
            ["./festival/bin/festival", "-i", "--pipe"],
            input=script.encode(),
            stderr=subprocess.STDOUT,
        ).decode()
    except subprocess.CalledProcessError as e:
        raise RuntimeError("festival実行エラー") from e
    print("=== festival出力 ===")
    print(output)
    return output


def extract_sexp(output: str) -> list[PhonemeInfo]:
    """Festival出力からS式部分を抽出し、PhonemeInfoリストに変換する"""
    print("=== S式抽出 ===")
    start = output.find('((("')
    if start == -1:
        raise RuntimeError("S式部分が見つかりません")
    paren = 0
    end = None
    for i, c in enumerate(output[start:], start=start):
        if c == "(":
            paren += 1
        elif c == ")":
            paren -= 1
            if paren == 0:
                end = i + 1
                break
    if end is None:
        raise RuntimeError("S式の括弧が閉じていません")
    sexp_text = output[start:end]
    print(sexp_text)
    print("=== S式パース ===")
    try:
        sexp = sexpdata.loads(sexp_text)
    except Exception as e:
        raise RuntimeError("sexpdata.loads失敗") from e

    infos: list[PhonemeInfo] = []
    for word_entry in sexp:
        witem = word_entry[0][0]
        word = witem.value() if isinstance(witem, sexpdata.Symbol) else str(witem)
        has_syl = False
        for syl_idx, syl_entry in enumerate(word_entry[1:], 1):
            if not (
                isinstance(syl_entry, list)
                and syl_entry
                and isinstance(syl_entry[0], list)
            ):
                continue
            key = syl_entry[0][0]
            if not (
                (isinstance(key, str) and key == "syl")
                or (isinstance(key, sexpdata.Symbol) and key.value() == "syl")
            ):
                continue
            has_syl = True
            syl_props = syl_entry[0][1]
            stress = next(
                (
                    p[1]
                    for p in syl_props
                    if isinstance(p, list)
                    and (
                        (isinstance(p[0], str) and p[0] == "stress")
                        or (
                            isinstance(p[0], sexpdata.Symbol)
                            and p[0].value() == "stress"
                        )
                    )
                ),
                None,
            )
            for ph_container in syl_entry[1:]:
                if not (
                    isinstance(ph_container, list)
                    and ph_container
                    and isinstance(ph_container[0], list)
                ):
                    continue
                psym = ph_container[0][0]
                phoneme = (
                    psym.value() if isinstance(psym, sexpdata.Symbol) else str(psym)
                )
                infos.append(
                    PhonemeInfo(
                        word=word,
                        syllable_index=syl_idx,
                        phoneme=phoneme,
                        stress=int(stress) if stress is not None else None,
                    )
                )
        if not has_syl:
            infos.append(
                PhonemeInfo(
                    word=word,
                    syllable_index=None,
                    phoneme=None,
                    stress=None,
                )
            )
    return infos


def print_phoneme_info(infos: list[PhonemeInfo]) -> None:
    """PhonemeInfoリストを見やすいJSONでprintする"""
    print(
        json.dumps([info.model_dump() for info in infos], ensure_ascii=False, indent=2)
    )


if __name__ == "__main__":
    main()
