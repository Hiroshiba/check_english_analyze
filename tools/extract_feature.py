"""
英語テキストから音素・シラブル・ストレス情報をjsonで出力するツール。

Usage:
    PYTHONPATH=. uv run python tools/extract_feature.py "internationalization"
    PYTHONPATH=. uv run python tools/extract_feature.py "hello, world!" --verbose
"""

import json

from pydantic import BaseModel

from tools.process_festival import PhonemeInfo as FestivalInfo
from tools.process_festival import festival as run_festival
from tools.process_phonemizer import PhonemeInfo as PhonemizerInfo
from tools.process_phonemizer import phonemizer_espeak


class UnifiedPhonemeInfo(BaseModel):
    """単語・シラブル・音素・ストレス・インデックス情報"""

    word: str
    word_index: int
    syllable_index: int
    phoneme: str
    phoneme_index: int
    stress: int


def main() -> None:
    """コマンドライン引数から実行するエントリポイント"""
    text, verbose = parse_args()
    result = extract_feature(text, verbose)
    print_phoneme_info(result)


def parse_args(args: list[str] | None = None) -> tuple[str, bool]:
    """コマンドライン引数からテキストとverboseを取得する"""
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("text", type=str, help="解析するテキスト")
    parser.add_argument(
        "--verbose", action="store_true", help="詳細なデバッグ出力をstderrに出す"
    )
    parsed = parser.parse_args(args)
    return parsed.text, parsed.verbose


def extract_feature(text: str, verbose: bool) -> list[UnifiedPhonemeInfo]:
    """英語テキストから音素・シラブル・ストレス情報を抽出しUnifiedPhonemeInfoリストで返す"""
    fest = run_festival(text, verbose)
    phnm = phonemizer_espeak(text, verbose)

    if not _check_consistency(fest, phnm):
        raise RuntimeError("festival/phonemizerの単語・インデックス情報が一致しません")

    result: list[UnifiedPhonemeInfo] = []
    for f, p in zip(fest, phnm, strict=False):
        result.append(
            UnifiedPhonemeInfo(
                word=f.word,
                word_index=f.word_index,
                syllable_index=f.syllable_index,
                phoneme=p.phoneme,
                phoneme_index=f.phoneme_index,
                stress=p.stress,
            )
        )
    return result


def print_phoneme_info(infos: list[UnifiedPhonemeInfo]) -> None:
    """UnifiedPhonemeInfoリストを見やすいJSONで標準出力する"""
    print(
        json.dumps(
            [info.model_dump() for info in infos],
            ensure_ascii=False,
            indent=2,
        )
    )


def _check_consistency(fest: list[FestivalInfo], phnm: list[PhonemizerInfo]) -> bool:
    """festival/phonemizerの単語・インデックス情報が一致しているか確認する"""
    if len(fest) != len(phnm):
        return False
    for f, p in zip(fest, phnm, strict=False):
        if (
            f.word != p.word
            or f.word_index != p.word_index
            or f.phoneme_index != p.phoneme_index
        ):
            return False
    return True


if __name__ == "__main__":
    main()
