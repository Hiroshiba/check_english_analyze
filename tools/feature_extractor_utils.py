"""特徴抽出処理に関連するユーティリティ関数群"""

from pathlib import Path

from tools.process_syllable import UnifiedPhonemeInfo
from utility.logger_utility import get_logger

logger = get_logger(Path(__file__))


def add_silence_phonemes(
    unified_infos: list[UnifiedPhonemeInfo],
) -> list[UnifiedPhonemeInfo]:
    """unified_infosの先頭・末尾に無音要素を追加する"""
    result = []

    result.append(
        UnifiedPhonemeInfo(
            word="",
            word_index=0,
            syllable_index=0,
            phoneme="",
            phoneme_index=0,
            stress=0,
        )
    )

    for info in unified_infos:
        result.append(
            UnifiedPhonemeInfo(
                word=info.word,
                word_index=info.word_index + 1,
                syllable_index=info.syllable_index + 1,
                phoneme=info.phoneme,
                phoneme_index=info.phoneme_index + 1,
                stress=info.stress,
            )
        )

    result.append(
        UnifiedPhonemeInfo(
            word="",
            word_index=result[-1].word_index + 1,
            syllable_index=result[-1].syllable_index + 1,
            phoneme="",
            phoneme_index=result[-1].phoneme_index + 1,
            stress=0,
        )
    )
    return result
