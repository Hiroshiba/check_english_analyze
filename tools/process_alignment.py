"""
英語音声ファイルとテキストファイルを元に音素アライメントのlabファイルを出力するツール。

Usage:
    PYTHONPATH=. uv run python tools/process_alignment.py --text-glob "tools/data/*.txt" --wav-glob "tools/data/*.wav" --output-dir ./hiho_aligned_output
    PYTHONPATH=. uv run python tools/process_alignment.py --text-glob "tools/data/*.txt" --wav-glob "tools/data/*.wav" --output-dir ./hiho_aligned_output --output-textgrid-dir ./hiho_aligned_output --verbose
"""

import tempfile
from pathlib import Path
from typing import Annotated

import typer

from tools.mfa_runner import (
    ensure_model_exists,
    prepare_corpus_dir,
    run_mfa_align,
    run_mfa_g2p,
    validate_mfa_command,
)
from tools.textgrid_parser import (
    LabEntry,
    copy_textgrid_files,
    parse_textgrid_file,
    write_lab_file,
)
from utility.file_utility import expand_glob_pattern
from utility.logger_utility import get_logger, logging_setting

logger = get_logger(Path(__file__))


def main(
    text_glob: Annotated[
        str,
        typer.Option(
            ...,
            help="テキストファイルのglobパターン（例: tools/data/*.txt）",
        ),
    ],
    wav_glob: Annotated[
        str,
        typer.Option(
            ...,
            help="音声ファイルのglobパターン（例: tools/data/*.wav）",
        ),
    ],
    output_dir: Annotated[
        Path,
        typer.Option(..., help="出力先ディレクトリ"),
    ],
    output_textgrid_dir: Annotated[
        Path | None,
        typer.Option(
            help="TextGridファイルの出力先ディレクトリ。指定しない場合は出力しない。"
        ),
    ] = None,
    verbose: Annotated[
        bool, typer.Option(help="詳細なデバッグ出力をstderrに出す")
    ] = False,
) -> None:
    """コマンドライン引数から実行するエントリポイント"""
    logging_setting(verbose)
    lab_dict = alignment(text_glob, wav_glob, output_textgrid_dir)
    write_lab_files(lab_dict, output_dir)


def alignment(
    text_glob: str,
    wav_glob: str,
    output_textgrid_dir: Path | None = None,
) -> dict[str, list[LabEntry]]:
    """アライメント処理本体。各ファイル名ごとにLabEntryリストを返す"""
    logger.debug(f"text_glob: {text_glob}")
    logger.debug(f"wav_glob: {wav_glob}")

    text_paths = expand_glob_pattern(text_glob, "テキストファイル")
    wav_paths = expand_glob_pattern(wav_glob, "音声ファイル")

    validate_mfa_command()
    validate_file_counts(text_paths, wav_paths)

    acoustic_model_name = "english_us_arpa"
    existing_dictionary_name = "english_us_arpa"
    g2p_model_name = "english_us_arpa"

    ensure_model_exists("acoustic", acoustic_model_name)
    ensure_model_exists("dictionary", existing_dictionary_name)
    ensure_model_exists("g2p", g2p_model_name)

    logger.debug("一時ディレクトリを作成")
    with tempfile.TemporaryDirectory() as temp_dir:
        temp_path = Path(temp_dir)
        corpus_dir = temp_path / "mfa_corpus"
        g2p_output_dictionary_path = temp_path / "g2p_dictionary.txt"
        textgrid_dir = temp_path / "textgrid_output"

        logger.debug(f"コーパスディレクトリ: {corpus_dir}")
        logger.debug(f"G2P辞書出力パス: {g2p_output_dictionary_path}")
        logger.debug(f"TextGrid出力ディレクトリ: {textgrid_dir}")

        prepare_corpus_dir(text_paths, wav_paths, corpus_dir)

        logger.info("G2P辞書を生成中...")
        g2p_result = run_mfa_g2p(
            corpus_dir,
            g2p_model_name,
            g2p_output_dictionary_path,
        )
        logger.debug(f"G2P処理完了: {g2p_result}")

        align_result = run_mfa_align(
            corpus_dir,
            str(g2p_output_dictionary_path),
            acoustic_model_name,
            textgrid_dir,
        )
        logger.debug("MFAアライメント完了")
        logger.debug(f"MFA出力: {align_result}")

        lab_dict: dict[str, list[LabEntry]] = {}
        for textgrid_file in textgrid_dir.glob("*.TextGrid"):
            logger.debug(f"TextGridファイル変換: {textgrid_file}")
            lab_entries = parse_textgrid_file(textgrid_file)
            lab_dict[textgrid_file.stem] = lab_entries
            logger.debug(f"phones: {[entry.phoneme for entry in lab_entries]}")

        if output_textgrid_dir is not None:
            copy_textgrid_files(textgrid_dir, output_textgrid_dir)

        return lab_dict


def write_lab_files(lab_dict: dict[str, list[LabEntry]], output_dir: Path) -> None:
    """LabEntryリストのdictをlabファイルとして出力する"""
    output_dir.mkdir(exist_ok=True)
    for stem, entries in lab_dict.items():
        lab_path = output_dir / f"{stem}.lab"
        write_lab_file(entries, lab_path)
        logger.debug(f"labファイル出力: {lab_path}")


def validate_file_counts(text_paths: list[Path], wav_paths: list[Path]) -> None:
    """テキストファイルと音声ファイルの数が一致するか検証し、不一致なら例外を投げる"""
    if len(text_paths) != len(wav_paths):
        raise ValueError(
            f"テキストファイルと音声ファイルの数が一致しません: text_paths={len(text_paths)}, wav_paths={len(wav_paths)}"
        )


if __name__ == "__main__":
    typer.run(main)
