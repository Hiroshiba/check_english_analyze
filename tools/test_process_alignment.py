import tempfile
from pathlib import Path

from syrupy.assertion import SnapshotAssertion

from tools.process_alignment import alignment


# TODO: multi_speaker=Trueのテストを追加する
def test_alignment_with_real_data(snapshot: SnapshotAssertion):
    """tools/data/*.txt, *.wavを使ってalignment関数の出力をスナップショットテスト"""
    text_glob = "tools/data/*.txt"
    wav_glob = "tools/data/*.wav"
    with tempfile.TemporaryDirectory() as temp_dir:
        result = alignment(
            text_glob, wav_glob, multi_speaker=False, output_textgrid_dir=Path(temp_dir)
        )
        assert result == snapshot
