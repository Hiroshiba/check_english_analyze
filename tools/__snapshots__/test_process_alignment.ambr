# serializer version: 1
# name: test_alignment_with_real_data
  dict({
    'Hello,-world': list([
      LabEntry(start=0.0, end=0.15, phoneme='(.)'),
      LabEntry(start=0.15, end=0.31, phoneme='h'),
      LabEntry(start=0.31, end=0.36, phoneme='ɛ'),
      LabEntry(start=0.36, end=0.46, phoneme='l'),
      LabEntry(start=0.46, end=0.83, phoneme='ow'),
      LabEntry(start=0.83, end=0.96, phoneme='(.)'),
      LabEntry(start=0.96, end=1.08, phoneme='w'),
      LabEntry(start=1.08, end=1.28, phoneme='ɝ'),
      LabEntry(start=1.28, end=1.32, phoneme='ɫ'),
      LabEntry(start=1.32, end=1.39, phoneme='d'),
      LabEntry(start=1.39, end=1.695057, phoneme='(.)'),
    ]),
    'libritts-6241_61946_000012_000002': list([
      LabEntry(start=0.0, end=0.06, phoneme='(.)'),
      LabEntry(start=0.06, end=0.14, phoneme='dʒ'),
      LabEntry(start=0.14, end=0.24, phoneme='i'),
      LabEntry(start=0.24, end=0.35, phoneme='ɑ'),
      LabEntry(start=0.35, end=0.38, phoneme='ɡ'),
      LabEntry(start=0.38, end=0.41, phoneme='ɹ'),
      LabEntry(start=0.41, end=0.45, phoneme='ə'),
      LabEntry(start=0.45, end=0.53, phoneme='f'),
      LabEntry(start=0.53, end=0.63, phoneme='ɚ'),
      LabEntry(start=0.63, end=0.67, phoneme='z'),
      LabEntry(start=0.67, end=0.7, phoneme='h'),
      LabEntry(start=0.7, end=0.73, phoneme='æ'),
      LabEntry(start=0.73, end=0.78, phoneme='v'),
      LabEntry(start=0.78, end=0.82, phoneme='dʲ'),
      LabEntry(start=0.82, end=0.87, phoneme='ɪ'),
      LabEntry(start=0.87, end=0.93, phoneme='v'),
      LabEntry(start=0.93, end=1.07, phoneme='aj'),
      LabEntry(start=1.07, end=1.1, phoneme='dʲ'),
      LabEntry(start=1.1, end=1.14, phoneme='ɪ'),
      LabEntry(start=1.14, end=1.18, phoneme='d'),
      LabEntry(start=1.18, end=1.25, phoneme='ɪ'),
      LabEntry(start=1.25, end=1.28, phoneme='t'),
      LabEntry(start=1.28, end=1.32, phoneme='ɪ'),
      LabEntry(start=1.32, end=1.38, phoneme='n'),
      LabEntry(start=1.38, end=1.43, phoneme='tʰ'),
      LabEntry(start=1.43, end=1.46, phoneme='ʉ'),
      LabEntry(start=1.46, end=1.6, phoneme='f'),
      LabEntry(start=1.6, end=1.67, phoneme='ɒ'),
      LabEntry(start=1.67, end=1.72, phoneme='ɹ'),
      LabEntry(start=1.72, end=1.84, phoneme='pʰ'),
      LabEntry(start=1.84, end=1.91, phoneme='ɑ'),
      LabEntry(start=1.91, end=2.02, phoneme='ɹ'),
      LabEntry(start=2.02, end=2.11, phoneme='t'),
      LabEntry(start=2.11, end=2.19, phoneme='s'),
      LabEntry(start=2.19, end=2.38, phoneme='(.)'),
      LabEntry(start=2.38, end=2.46, phoneme='æ'),
      LabEntry(start=2.46, end=2.49, phoneme='n'),
      LabEntry(start=2.49, end=2.54, phoneme='d'),
      LabEntry(start=2.54, end=2.58, phoneme='w'),
      LabEntry(start=2.58, end=2.62, phoneme='i'),
      LabEntry(start=2.62, end=2.67, phoneme='h'),
      LabEntry(start=2.67, end=2.72, phoneme='æ'),
      LabEntry(start=2.72, end=2.75, phoneme='d'),
      LabEntry(start=2.75, end=2.78, phoneme='t'),
      LabEntry(start=2.78, end=2.81, phoneme='ə'),
      LabEntry(start=2.81, end=2.95, phoneme='k'),
      LabEntry(start=2.95, end=2.99, phoneme='ɹ'),
      LabEntry(start=2.99, end=3.11, phoneme='ɑː'),
      LabEntry(start=3.11, end=3.22, phoneme='s'),
      LabEntry(start=3.22, end=3.34, phoneme='θ'),
      LabEntry(start=3.34, end=3.46, phoneme='s'),
      LabEntry(start=3.46, end=3.61, phoneme='aw'),
      LabEntry(start=3.61, end=3.71, phoneme='θ'),
      LabEntry(start=3.71, end=3.78, phoneme='w'),
      LabEntry(start=3.78, end=3.84, phoneme='ɛ'),
      LabEntry(start=3.84, end=3.93, phoneme='s'),
      LabEntry(start=3.93, end=3.96, phoneme='t'),
      LabEntry(start=3.96, end=4.07, phoneme='kʷ'),
      LabEntry(start=4.07, end=4.14, phoneme='ɒ'),
      LabEntry(start=4.14, end=4.19, phoneme='ɹ'),
      LabEntry(start=4.19, end=4.23, phoneme='t'),
      LabEntry(start=4.23, end=4.38, phoneme='ɚ'),
      LabEntry(start=4.38, end=4.69, phoneme='(.)'),
      LabEntry(start=4.69, end=4.8, phoneme='w'),
      LabEntry(start=4.8, end=4.85, phoneme='ɪ'),
      LabEntry(start=4.85, end=4.96, phoneme='tʃ'),
      LabEntry(start=4.96, end=5.09, phoneme='ɪ'),
      LabEntry(start=5.09, end=5.13, phoneme='n'),
      LabEntry(start=5.13, end=5.2, phoneme='ð'),
      LabEntry(start=5.2, end=5.23, phoneme='v'),
      LabEntry(start=5.23, end=5.27, phoneme='ɚ'),
      LabEntry(start=5.27, end=5.34, phoneme='ɲ'),
      LabEntry(start=5.34, end=5.48, phoneme='æ'),
      LabEntry(start=5.48, end=5.57, phoneme='c'),
      LabEntry(start=5.57, end=5.61, phoneme='ə'),
      LabEntry(start=5.61, end=5.64, phoneme='l'),
      LabEntry(start=5.64, end=5.75, phoneme='ɚ'),
      LabEntry(start=5.75, end=5.83, phoneme='ɪ'),
      LabEntry(start=5.83, end=5.88, phoneme='z'),
      LabEntry(start=5.88, end=5.98, phoneme='kʰ'),
      LabEntry(start=5.98, end=6.05, phoneme='ɒː'),
      LabEntry(start=6.05, end=6.23, phoneme='ɫ'),
      LabEntry(start=6.23, end=6.29, phoneme='d'),
      LabEntry(start=6.29, end=6.54, phoneme='(.)'),
      LabEntry(start=6.54, end=6.71, phoneme='s'),
      LabEntry(start=6.71, end=6.8, phoneme='ʉ'),
      LabEntry(start=6.8, end=6.86, phoneme='d'),
      LabEntry(start=6.86, end=6.9, phoneme='v'),
      LabEntry(start=6.9, end=7.02, phoneme='ɛ'),
      LabEntry(start=7.02, end=7.09, phoneme='s'),
      LabEntry(start=7.09, end=7.12, phoneme='t'),
      LabEntry(start=7.12, end=7.13, phoneme='ɹ'),
      LabEntry(start=7.13, end=7.29, phoneme='fʲ'),
      LabEntry(start=7.29, end=7.32, phoneme='i'),
      LabEntry(start=7.32, end=7.36, phoneme='ɒ'),
      LabEntry(start=7.36, end=7.41, phoneme='ɹ'),
      LabEntry(start=7.41, end=7.44, phoneme='d'),
      LabEntry(start=7.44, end=7.5, phoneme='ə'),
      LabEntry(start=7.5, end=7.55, phoneme='ŋ'),
      LabEntry(start=7.55, end=7.59, phoneme='ɡ'),
      LabEntry(start=7.59, end=7.68, phoneme='ɹ'),
      LabEntry(start=7.68, end=7.9, phoneme='(.)'),
    ]),
    'librittsr-6241_61946_000012_000002': list([
      LabEntry(start=0.0, end=0.08, phoneme='(.)'),
      LabEntry(start=0.08, end=0.18, phoneme='dʒ'),
      LabEntry(start=0.18, end=0.27, phoneme='i'),
      LabEntry(start=0.27, end=0.37, phoneme='ɑ'),
      LabEntry(start=0.37, end=0.4, phoneme='ɡ'),
      LabEntry(start=0.4, end=0.43, phoneme='ɹ'),
      LabEntry(start=0.43, end=0.47, phoneme='ə'),
      LabEntry(start=0.47, end=0.55, phoneme='f'),
      LabEntry(start=0.55, end=0.65, phoneme='ɚ'),
      LabEntry(start=0.65, end=0.69, phoneme='z'),
      LabEntry(start=0.69, end=0.72, phoneme='h'),
      LabEntry(start=0.72, end=0.75, phoneme='æ'),
      LabEntry(start=0.75, end=0.78, phoneme='v'),
      LabEntry(start=0.78, end=0.85, phoneme='dʲ'),
      LabEntry(start=0.85, end=0.88, phoneme='ɪ'),
      LabEntry(start=0.88, end=0.95, phoneme='v'),
      LabEntry(start=0.95, end=1.08, phoneme='aj'),
      LabEntry(start=1.08, end=1.11, phoneme='dʲ'),
      LabEntry(start=1.11, end=1.14, phoneme='ɪ'),
      LabEntry(start=1.14, end=1.2, phoneme='d'),
      LabEntry(start=1.2, end=1.26, phoneme='ɪ'),
      LabEntry(start=1.26, end=1.29, phoneme='t'),
      LabEntry(start=1.29, end=1.34, phoneme='ɪ'),
      LabEntry(start=1.34, end=1.4, phoneme='n'),
      LabEntry(start=1.4, end=1.45, phoneme='tʰ'),
      LabEntry(start=1.45, end=1.48, phoneme='ʉ'),
      LabEntry(start=1.48, end=1.62, phoneme='f'),
      LabEntry(start=1.62, end=1.69, phoneme='ɒ'),
      LabEntry(start=1.69, end=1.75, phoneme='ɹ'),
      LabEntry(start=1.75, end=1.87, phoneme='pʰ'),
      LabEntry(start=1.87, end=1.93, phoneme='ɑ'),
      LabEntry(start=1.93, end=2.04, phoneme='ɹ'),
      LabEntry(start=2.04, end=2.13, phoneme='t'),
      LabEntry(start=2.13, end=2.2, phoneme='s'),
      LabEntry(start=2.2, end=2.37, phoneme='(.)'),
      LabEntry(start=2.37, end=2.48, phoneme='æ'),
      LabEntry(start=2.48, end=2.52, phoneme='n'),
      LabEntry(start=2.52, end=2.55, phoneme='d'),
      LabEntry(start=2.55, end=2.59, phoneme='w'),
      LabEntry(start=2.59, end=2.63, phoneme='i'),
      LabEntry(start=2.63, end=2.68, phoneme='h'),
      LabEntry(start=2.68, end=2.73, phoneme='æ'),
      LabEntry(start=2.73, end=2.76, phoneme='d'),
      LabEntry(start=2.76, end=2.79, phoneme='t'),
      LabEntry(start=2.79, end=2.84, phoneme='ə'),
      LabEntry(start=2.84, end=2.97, phoneme='k'),
      LabEntry(start=2.97, end=3.02, phoneme='ɹ'),
      LabEntry(start=3.02, end=3.14, phoneme='ɑː'),
      LabEntry(start=3.14, end=3.23, phoneme='s'),
      LabEntry(start=3.23, end=3.37, phoneme='θ'),
      LabEntry(start=3.37, end=3.48, phoneme='s'),
      LabEntry(start=3.48, end=3.63, phoneme='aw'),
      LabEntry(start=3.63, end=3.72, phoneme='θ'),
      LabEntry(start=3.72, end=3.8, phoneme='w'),
      LabEntry(start=3.8, end=3.86, phoneme='ɛ'),
      LabEntry(start=3.86, end=3.96, phoneme='s'),
      LabEntry(start=3.96, end=3.99, phoneme='t'),
      LabEntry(start=3.99, end=4.09, phoneme='kʷ'),
      LabEntry(start=4.09, end=4.16, phoneme='ɒ'),
      LabEntry(start=4.16, end=4.2, phoneme='ɹ'),
      LabEntry(start=4.2, end=4.24, phoneme='t'),
      LabEntry(start=4.24, end=4.44, phoneme='ɚ'),
      LabEntry(start=4.44, end=4.71, phoneme='(.)'),
      LabEntry(start=4.71, end=4.81, phoneme='w'),
      LabEntry(start=4.81, end=4.86, phoneme='ɪ'),
      LabEntry(start=4.86, end=4.97, phoneme='tʃ'),
      LabEntry(start=4.97, end=5.11, phoneme='ɪ'),
      LabEntry(start=5.11, end=5.16, phoneme='n'),
      LabEntry(start=5.16, end=5.22, phoneme='ð'),
      LabEntry(start=5.22, end=5.26, phoneme='v'),
      LabEntry(start=5.26, end=5.29, phoneme='ɚ'),
      LabEntry(start=5.29, end=5.36, phoneme='ɲ'),
      LabEntry(start=5.36, end=5.5, phoneme='æ'),
      LabEntry(start=5.5, end=5.6, phoneme='c'),
      LabEntry(start=5.6, end=5.63, phoneme='ə'),
      LabEntry(start=5.63, end=5.66, phoneme='l'),
      LabEntry(start=5.66, end=5.76, phoneme='ɚ'),
      LabEntry(start=5.76, end=5.85, phoneme='ɪ'),
      LabEntry(start=5.85, end=5.91, phoneme='z'),
      LabEntry(start=5.91, end=6.0, phoneme='kʰ'),
      LabEntry(start=6.0, end=6.12, phoneme='ɒ'),
      LabEntry(start=6.12, end=6.25, phoneme='ɫ'),
      LabEntry(start=6.25, end=6.31, phoneme='d'),
      LabEntry(start=6.31, end=6.57, phoneme='(.)'),
      LabEntry(start=6.57, end=6.73, phoneme='s'),
      LabEntry(start=6.73, end=6.81, phoneme='ʉ'),
      LabEntry(start=6.81, end=6.88, phoneme='d'),
      LabEntry(start=6.88, end=6.91, phoneme='v'),
      LabEntry(start=6.91, end=7.04, phoneme='ɛ'),
      LabEntry(start=7.04, end=7.11, phoneme='s'),
      LabEntry(start=7.11, end=7.14, phoneme='t'),
      LabEntry(start=7.14, end=7.17, phoneme='ɹ'),
      LabEntry(start=7.17, end=7.31, phoneme='fʲ'),
      LabEntry(start=7.31, end=7.34, phoneme='i'),
      LabEntry(start=7.34, end=7.37, phoneme='ɒ'),
      LabEntry(start=7.37, end=7.42, phoneme='ɹ'),
      LabEntry(start=7.42, end=7.46, phoneme='d'),
      LabEntry(start=7.46, end=7.51, phoneme='ə'),
      LabEntry(start=7.51, end=7.57, phoneme='ŋ'),
      LabEntry(start=7.57, end=7.61, phoneme='ɡ'),
      LabEntry(start=7.61, end=7.71, phoneme='ɹ'),
      LabEntry(start=7.71, end=7.88, phoneme='(.)'),
    ]),
  })
# ---
